_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,229 @@
+``` 
+
+
+### Special case when the number of variants which has payload is only 1.
+
+Take the types below for example:
+
+```reasonml
+type list = 
+    | Nil
+    | Cons (int * list);
+```
+
+Since the number of variants which has payload is only 1, the compiler does not need add `TAG` when we destruct the data for pattern matching, so the code below:
+
+```reasonml
+let u = Cons(1,Nil)
+```
+Will generate such JS output:
+
+```js
+var u = {_0: 1, _1 : /*Nil*/ 0 }; // No TAG data.
+```
+
+### Specialized for immutable list
+
+The `list` type is a built-in type, its type definition is similar to this :
+
+```reasonml
+type t ('a) = 
+    | []
+    | (::) ('a * t ('a))
+```
+
+Without any customization,  it will generate js objects with indexes like `_0`, `_1`, since list is so pervasive,
+we provide some special treatment so that 
+```reasonml
+let u = [0,1,2,3]
+```
+Will generate js code as below:
+```js
+var u = {hd : 0, {hd : 1, {hd: 2, {hd :3 , tl : /*[]*/0 }}}}
+```
+
+This is a minor change, we changed the name of `_0` to `hd` and `_1` to `tl`.
+
+### IC friendliness
+
+For types whose number of variants which has payload is 1, it will be in a perfect IC position. 
+The number of variants which does not carry payload will not affect IC, since the pattern match will do a split first. 
+
+For types whose number of variants which has the same number of payloads, it will also be in a perfect IC position, like the red-black-tree example above. 
+
+For other cases, it will hit a polymorphic IC in the V8 jit compiler, this is not the fastest running case.
+
+Note that user can always tweak the variant layout to make it IC friendly, for example, it can always introduce one level of indirection to make all variants share the same number of payloads:
+
+```reasonml
+type t = 
+    | A0 of a0 // 1 paylaod
+    | A1 of a1 // 1 payload
+    | A2 of a2 // 1 payload
+    | C0
+    | C1
+    | C2 // This will not affect IC 
+```
+
+### Variant in debug mode
+
+Note we only generate constructor names in comments for debugging,  when constructor names are attached to the data, it will be more useful for debugging. When debug mode is activated using `-bs-g`, 
+
+The generated code will be changed from below 
+```js
+var v0  = {TAG : 0/*Black*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+```
+to
+
+```js
+var v0  = {TAG : 0, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0, [Symbol.for("name")]: "Black"};
+```
+
+## Polymorphic-variant (internal)
+
+Polymorphic variant allows users to use the types without declaring it first:
+
+```reasonml
+let u = 3 -> `hello
+```
+
+It will generate
+```js
+var u = {HASH : MAGIC_NUMBER, VAL: 3 }
+```
+The field of `HASH` is the hash of name `"hello"`, while the `VAL` is the payload
+
+### IC friendliness
+
+Polymorphic variant is always in a perfect IC position, the compiler can ensure all generated records are of the same shape. This is due to that the payload is not unpacked, it is always just one payload
```

Comment:
Polymorphic variants are always in a perfect position for IC. The compiler can ensure all generated records...
This is because the payload is not unpacked; it is always just one payload.